libraries/custom_appbase/include/eosio/chain/application.hpp

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <appbase/application_base.hpp>
+#include <appbase/execution_priority_queue.hpp>
+
+/*
+ * Custmomize appbase to support two-queue exector.
+*/
+namespace appbase { 
+
+enum class exec_window {
+   read,  // the window during which operations from read_only_trx_safe queue
+          // can be executed in app thread in parallel with read-only transactions
+          // in read-only transaction excuting threads.
+   write, // the window during which operations from both general and
+          // read_only_trx_safe queues can be executed in app thread,
+          // while read-only transactions are not executed in read-only
+          // transaction excuting threads.
+};
+
+enum class exec_queue {
+   read_only_trx_safe, // the queue storing operations which are safe to execute
+                       // on app thread in parallel with read-only transactions
+                       // in read-only transaction excuting threads.
+   general             // the queue storing operations which can be only executed
+                       // on the app thread while read-only transactions are
+                       // not being executed in read-only threads
+};
+
+class two_queue_executor {
+public:
+
+   template <typename Func>
+   auto post( int priority, exec_queue q, Func&& func ) {
+      if ( q == exec_queue::general )
+         return boost::asio::post(io_serv_, general_queue_.wrap(priority, --order_, std::forward<Func>(func)));
+      else
+         return boost::asio::post(io_serv_, read_only_trx_safe_queue_.wrap(priority, --order_, std::forward<Func>(func)));
+   }
+
+   // Legacy and deprecated. To be removed after cleaning up its uses in base appbase
+   template <typename Func>
+   auto post( int priority, Func&& func ) {
+      // safer to use general queue for unknown type of operation since operations
+      // from general queue are not executed in parallel with read-only transactions
+      return boost::asio::post(io_serv_, general_queue_.wrap(priority, --order_, std::forward<Func>(func)));
+   }
+
+   boost::asio::io_service& get_io_service() { return io_serv_; }
+
+   bool execute_highest() {
+      if ( exec_window_ == exec_window::write ) {
+         if( !general_queue_.empty() && ( read_only_trx_safe_queue_.empty() || *read_only_trx_safe_queue_.top() < *general_queue_.top()) )  {
+            // general_queue_'s top function's priority greater than read_only_trx_safe_queue_'s top function's, or general_queue_ empty
+            general_queue_.execute_highest();
+         } else if( !read_only_trx_safe_queue_.empty() ) {
+            read_only_trx_safe_queue_.execute_highest();
+         }
+         return !read_only_trx_safe_queue_.empty() || !general_queue_.empty();
+      } else {
+         return read_only_trx_safe_queue_.execute_highest();
+      }
+   }
```

Comment:
I think this is suboptimal. 
Instead of having 2 queues, each based on `std::priority_queue`, I think it would be better to have one queue based on a `boost::multi_index_container` with 2 ordered_non_unique indices, one with a predicate just based on priority (for readwrite execution), and one with a predicate based on (readonly, priority) which would have all the readonly first (for readonly execution).
The readwrite execution, in particular, would likely be more efficient.

Reply:
Thanks for the suggestion. We will look at further improvements in future tasks.